### ISSUE TITLE
[Issue]: Saving of large PNGs got slower recently

### ISSUE THREAD

**Issue by GodEmperor785** [vladmandic/automatic#2305]
### Issue Description

I noticed that recently saving or larger PNG images got slower than it used to be. Some weeks ago I didn't notice any significant time to save bigger PNGs. But recently saving them takes a really noticeable time. This was especially noticeable on saving grids as PNG, larger grids took really long time to save. Or when generating several larger images - there were noticeable gaps in time between generations for saving images.

I checked the code a bit and found that line added in this relatively recent commit added max compression level for PNGs (compression_level=9 and optimize=True):
https://github.com/vladmandic/automatic/commit/cefb5959e33fc04cb679c7f12d78dcb36a93586d

Long time to save PNGs can be reproduced easily by "upscaling" some image in "process image" tab but without actually choosing upscaler model (so it remains "None") so input image simply gets saved with no size change

Without compression_level=9 and optimize=True images save much faster and they aren't much bigger in size from my limited tests (optimize=True forces compression_level=9 regardless of compression_level parameter value according to documentation mnentioned in commit so in tests I removed both parameters):
with compression currently on master:
upscaling 1152x1728 without model - file size is 2.1MB, and saving time is around 4-4.5s
with default compression (both parameters removed):
upscaling 1152x1728 without model - file size is 2.17MB, and saving time is around 0.6-0.7s

Time for the upscaling images was measured by adding 'shared.log.debug("sth")' before and after relevant line in code (modules/images.py line 465 on master) and calculating diff in time. Tested on images originally saved using different compression levels and saving time is consistent, output image size is given for the same image to be able to compare

Form the results it looks like there is little difference in result file size but over 6x difference in saving time

Should compression level be prioritized so much over speed during saving PNGs while there is a very small gain in file size? Maybe default/lower compression level should be used to speed things up a bit or perhaps a new setting could be added to control compression level of PNGs?

Not sure if any logs in "Relevant log output" are really needed in this case? If there are any let me know and I will add them

### Version Platform Description

19:04:16-680417 INFO     Starting SD.Next
19:04:16-683417 INFO     Python 3.10.6 on Windows
19:04:16-787416 INFO     Version: app=sd.next updated=2023-10-01 hash=c6226605
                         url=https://github.com/vladmandic/automatic.git/tree/master
19:04:17-018932 INFO     Latest published version: e5f8b7f0a462a3b1cc03b2a76ea50509138ba7d3 2023-10-07T17:19:56Z
19:04:17-031432 INFO     Platform: arch=AMD64 cpu=Intel64 Family 6 Model 183 Stepping 1, GenuineIntel system=Windows
                         release=Windows-10-10.0.19045-SP0 python=3.10.6
19:04:17-035432 DEBUG    Setting environment tuning
19:04:17-036432 DEBUG    Torch overrides: cuda=False rocm=False ipex=False diml=False openvino=False
19:04:17-038432 DEBUG    Torch allowed: cuda=True rocm=True ipex=True diml=True openvino=True
19:04:17-041932 INFO     nVidia CUDA toolkit detected
19:04:17-130431 WARNING  Modified files: ['modules/images.py']
19:04:17-175431 DEBUG    Repository update time: Sun Oct  1 17:25:49 2023
19:04:17-177931 DEBUG    Previous setup time: Tue Oct  3 17:00:29 2023
19:04:17-179931 INFO     Extensions: disabled=[]
19:04:17-180932 INFO     Extensions: enabled=['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora',
                         'multidiffusion-upscaler-for-automatic1111', 'ScuNET', 'sd-dynamic-thresholding',
                         'sd-extension-system-info', 'sd-webui-agent-scheduler', 'sd-webui-controlnet',
                         'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR']
                         extensions-builtin
19:04:17-186931 INFO     Extensions: enabled=['adetailer', 'sd-webui-freeu', 'sd-webui-lora-block-weight',
                         'sd-webui-openpose-editor', 'sd-webui-stablesr', 'sd-webui-vectorscope-cc', 'sd_webui_SAG',
                         'stable-diffusion-webui-anti-burn', 'Stable-Diffusion-Webui-Civitai-Helper',
                         'ultimate-upscale-for-automatic1111'] extensions
19:04:17-191431 DEBUG    Latest extensions time: Tue Oct  3 17:00:22 2023
19:04:17-192431 DEBUG    Timestamps: version:1696173949 setup:1696345229 extension:1696345222
19:04:17-193932 INFO     No changes detected: Quick launch active
19:04:17-194431 INFO     Verifying requirements
19:04:17-201931 INFO     Verifying packages
19:04:17-203431 INFO     Extensions: disabled=[]
19:04:17-204431 INFO     Extensions: enabled=['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora',
                         'multidiffusion-upscaler-for-automatic1111', 'ScuNET', 'sd-dynamic-thresholding',
                         'sd-extension-system-info', 'sd-webui-agent-scheduler', 'sd-webui-controlnet',
                         'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR']
                         extensions-builtin
19:04:17-214431 INFO     Extensions: enabled=['adetailer', 'sd-webui-freeu', 'sd-webui-lora-block-weight',
                         'sd-webui-openpose-editor', 'sd-webui-stablesr', 'sd-webui-vectorscope-cc', 'sd_webui_SAG',
                         'stable-diffusion-webui-anti-burn', 'Stable-Diffusion-Webui-Civitai-Helper',
                         'ultimate-upscale-for-automatic1111'] extensions
19:04:17-228106 INFO     Extension preload: {'extensions-builtin': 0.01, 'extensions': 0.0}
19:04:17-230606 INFO     Command line args: ['--autolaunch', '--debug'] autolaunch=True debug=True
19:04:22-302425 DEBUG    Loaded packages: torch=2.0.1+cu118 diffusers=0.20.2 gradio=3.43.2
19:04:22-639856 INFO     Engine: backend=Backend.ORIGINAL compute=cuda mode=no_grad device=cuda
19:04:22-684856 INFO     Device: device=NVIDIA GeForce RTX 3090 n=1 arch=compute_37 cap=(8, 6) cuda=11.8 cudnn=8700
                         driver=537.42

### Backend

Original

### Model

SD 1.5

### Acknowledgements

- [X] I have read the above and searched for existing issues
- [X] I confirm that this is classified correctly and its not an extension issue

**Comment by vladmandic** [comment]
that's too much of an impact, i'll remove optimize=True for PNG files.

**Comment by Aptronymist** [comment]
> ### Issue Description
> I noticed that recently saving or larger PNG images got slower than it used to be. Some weeks ago I didn't notice any significant time to save bigger PNGs. But recently saving them takes a really noticeable time. This was especially noticeable on saving grids as PNG, larger grids took really long time to save. Or when generating several larger images - there were noticeable gaps in time between generations for saving images.
> 
> I checked the code a bit and found that line added in this relatively recent commit added max compression level for PNGs (compression_level=9 and optimize=True): [cefb595](https://github.com/vladmandic/automatic/commit/cefb5959e33fc04cb679c7f12d78dcb36a93586d)
> 
> Long time to save PNGs can be reproduced easily by "upscaling" some image in "process image" tab but without actually choosing upscaler model (so it remains "None") so input image simply gets saved with no size change
> 
> Without compression_level=9 and optimize=True images save much faster and they aren't much bigger in size from my limited tests (optimize=True forces compression_level=9 regardless of compression_level parameter value according to documentation mnentioned in commit so in tests I removed both parameters): with compression currently on master: upscaling 1152x1728 without model - file size is 2.1MB, and saving time is around 4-4.5s with default compression (both parameters removed): upscaling 1152x1728 without model - file size is 2.17MB, and saving time is around 0.6-0.7s
> 
> Time for the upscaling images was measured by adding 'shared.log.debug("sth")' before and after relevant line in code (modules/images.py line 465 on master) and calculating diff in time. Tested on images originally saved using different compression levels and saving time is consistent, output image size is given for the same image to be able to compare
> 
> Form the results it looks like there is little difference in result file size but over 6x difference in saving time
> 
> Should compression level be prioritized so much over speed during saving PNGs while there is a very small gain in file size? Maybe default/lower compression level should be used to speed things up a bit or perhaps a new setting could be added to control compression level of PNGs?
> 
> Not sure if any logs in "Relevant log output" are really needed in this case? If there are any let me know and I will add them
> 
> ### Version Platform Description
> 19:04:16-680417 INFO Starting SD.Next 19:04:16-683417 INFO Python 3.10.6 on Windows 19:04:16-787416 INFO Version: app=sd.next updated=2023-10-01 hash=c6226605 url=[vladmandic/automatic.git@`master`](https://github.com/vladmandic/automatic.git/tree/master?rgh-link-date=2023-10-13T17%3A26%3A33Z) 19:04:17-018932 INFO Latest published version: [e5f8b7f](https://github.com/vladmandic/automatic/commit/e5f8b7f0a462a3b1cc03b2a76ea50509138ba7d3) 2023-10-07T17:19:56Z 19:04:17-031432 INFO Platform: arch=AMD64 cpu=Intel64 Family 6 Model 183 Stepping 1, GenuineIntel system=Windows release=Windows-10-10.0.19045-SP0 python=3.10.6 19:04:17-035432 DEBUG Setting environment tuning 19:04:17-036432 DEBUG Torch overrides: cuda=False rocm=False ipex=False diml=False openvino=False 19:04:17-038432 DEBUG Torch allowed: cuda=True rocm=True ipex=True diml=True openvino=True 19:04:17-041932 INFO nVidia CUDA toolkit detected 19:04:17-130431 WARNING Modified files: ['modules/images.py'] 19:04:17-175431 DEBUG Repository update time: Sun Oct 1 17:25:49 2023 19:04:17-177931 DEBUG Previous setup time: Tue Oct 3 17:00:29 2023 19:04:17-179931 INFO Extensions: disabled=[] 19:04:17-180932 INFO Extensions: enabled=['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora', 'multidiffusion-upscaler-for-automatic1111', 'ScuNET', 'sd-dynamic-thresholding', 'sd-extension-system-info', 'sd-webui-agent-scheduler', 'sd-webui-controlnet', 'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR'] extensions-builtin 19:04:17-186931 INFO Extensions: enabled=['adetailer', 'sd-webui-freeu', 'sd-webui-lora-block-weight', 'sd-webui-openpose-editor', 'sd-webui-stablesr', 'sd-webui-vectorscope-cc', 'sd_webui_SAG', 'stable-diffusion-webui-anti-burn', 'Stable-Diffusion-Webui-Civitai-Helper', 'ultimate-upscale-for-automatic1111'] extensions 19:04:17-191431 DEBUG Latest extensions time: Tue Oct 3 17:00:22 2023 19:04:17-192431 DEBUG Timestamps: version:1696173949 setup:1696345229 extension:1696345222 19:04:17-193932 INFO No changes detected: Quick launch active 19:04:17-194431 INFO Verifying requirements 19:04:17-201931 INFO Verifying packages 19:04:17-203431 INFO Extensions: disabled=[] 19:04:17-204431 INFO Extensions: enabled=['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora', 'multidiffusion-upscaler-for-automatic1111', 'ScuNET', 'sd-dynamic-thresholding', 'sd-extension-system-info', 'sd-webui-agent-scheduler', 'sd-webui-controlnet', 'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR'] extensions-builtin 19:04:17-214431 INFO Extensions: enabled=['adetailer', 'sd-webui-freeu', 'sd-webui-lora-block-weight', 'sd-webui-openpose-editor', 'sd-webui-stablesr', 'sd-webui-vectorscope-cc', 'sd_webui_SAG', 'stable-diffusion-webui-anti-burn', 'Stable-Diffusion-Webui-Civitai-Helper', 'ultimate-upscale-for-automatic1111'] extensions 19:04:17-228106 INFO Extension preload: {'extensions-builtin': 0.01, 'extensions': 0.0} 19:04:17-230606 INFO Command line args: ['--autolaunch', '--debug'] autolaunch=True debug=True 19:04:22-302425 DEBUG Loaded packages: torch=2.0.1+cu118 diffusers=0.20.2 gradio=3.43.2 19:04:22-639856 INFO Engine: backend=Backend.ORIGINAL compute=cuda mode=no_grad device=cuda 19:04:22-684856 INFO Device: device=NVIDIA GeForce RTX 3090 n=1 arch=compute_37 cap=(8, 6) cuda=11.8 cudnn=8700 driver=537.42
> 
> ### Relevant log output
> _No response_
> 
> ### Backend
> Original
> 
> ### Model
> SD 1.5
> 
> ### Acknowledgements
> * [x]  I have read the above and searched for existing issues
> * [x]  I confirm that this is classified correctly and its not an extension issue

Good catch, I did not do any testing of larger image saving or slowdowns, it was just intended to correctly optimize them as is being done with jpegs, since that was wrong.

**Comment by GodEmperor785** [comment]
I made a quick test with a script to check exactly how each compression_level behaves and here are results:
compress_level=0 - time 0.07025552999994034 s | avg size =  5.698886871337891 MB
compress_level=1 - time 0.08770548000011331 s | avg size =  2.3558088302612306 MB
compress_level=2 - time 0.10568544000007023 s | avg size =  2.2939175605773925 MB
compress_level=3 - time 0.1724531199998637 s | avg size =  2.2322935104370116 MB
compress_level=4 - time 0.19454377999991265 s | avg size =  2.089144992828369 MB
compress_level=5 - time 0.3062161599999854 s | avg size =  2.0748318672180175 MB
compress_level=6 - time 0.6383302299999741 s | avg size =  2.050089454650879 MB
compress_level=7 - time 0.9476251500001126 s | avg size =  2.0289626121520996 MB
compress_level=8 - time 3.1463240700001736 s | avg size =  1.9855969429016114 MB
compress_level=9 - time 4.760710160000054 s | avg size =  1.9729284286499023 MB
compress_level=0 + optimize=True - time 4.797175669999978 s | avg size =  1.9677529335021973 MB

This was tested by loading a random generated image (all generated by me using this SD.Next in recent days using several different SD 1.5 models, all of size 1152x1728 for this test) and saving with different compression levels, it was repeated 10 times and result values are averages. PIL version should be the same as in SD.Next (9.5.0 as in requirements.txt)

It is visible that compression_level 9 and 8 are much slower and they don't offer much file size reduction and file saving still takes long time. 

I saw that on dev branch compress_level was changed to 8 which doesn't improve situation by much: https://github.com/vladmandic/automatic/commit/74fb8fba3e945acc0f688e19fde96c11ce6f2715

Wouldn't it be better to use default compress_level (6 according to documentation) or even 5 for bigger gain in speed?
